server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,757 @@
+    // miss when devicePath contains *
+    for (String devicePath : devicePaths) {
+      if (devicePath.contains("*")) {
+        return storageGroupToDeviceMap;
+      }
+    }
+    // first try to hit storage group in fast-fail way
+    boolean firstTryResult =
+        getStorageGroupToDeviceMap(devicePaths, storageGroupToDeviceMap, new ArrayList<>(), true);
+    if (!firstTryResult) {
+      try {
+        // when local cache not have, then try to fetch all storage group from config node
+        getAllStorageGroupFromConfigNode();
+        // second try to hit storage group with failed devices;
+        List<String> failedDevices = new ArrayList<>();
+        boolean secondTryResult =
+            getStorageGroupToDeviceMap(devicePaths, storageGroupToDeviceMap, failedDevices, false);
+        if (!secondTryResult && isAutoCreate) {
+          // try to auto create storage group of failed device
+          Set<String> storageGroupNamesNeedCreated = new HashSet<>();
+          for (String devicePath : failedDevices) {
+            PartialPath storageGroupNameNeedCreated =
+                MetaUtils.getStorageGroupPathByLevel(
+                    new PartialPath(devicePath), config.getDefaultStorageGroupLevel());
+            storageGroupNamesNeedCreated.add(storageGroupNameNeedCreated.getFullPath());
+          }
+          createNotExistedStorageGroup(storageGroupNamesNeedCreated);
+          // third try to hit cache
+          boolean thirdTryResult =
+              getStorageGroupToDeviceMap(
+                  devicePaths, storageGroupToDeviceMap, new ArrayList<>(), false);
+          if (!thirdTryResult) {
+            throw new StatementAnalyzeException("Failed to get Storage Group Map in three try.");
+          }
+        }
+      } catch (TException | MetadataException | IOException e) {
+        throw new StatementAnalyzeException(
+            "An error occurred when executing getStorageGroupToDevice():" + e.getMessage());
+      }
+    }
+    return storageGroupToDeviceMap;
+  }
+
+  /**
+   * get storage group to device map in three trys
+   *
+   * @param devicePaths the devices that need to match
+   * @param isAutoCreate whether auto create storage group when device miss
+   * @return device to storage group map
+   */
+  public Map<String, String> getDeviceToStorageGroup(
+      List<String> devicePaths, boolean isAutoCreate) {
+    Map<String, String> deviceToStorageGroupMap = new HashMap<>();
+    // miss when devicePath contains *
+    for (String devicePath : devicePaths) {
+      if (devicePath.contains("*")) {
+        return deviceToStorageGroupMap;
+      }
+    }
+    // first try to hit storage group in fast-fail way
+    boolean firstTryResult =
+        getDeviceToStorageGroupMap(devicePaths, deviceToStorageGroupMap, new ArrayList<>(), true);
+    if (!firstTryResult) {
+      try {
+        // when local cache not have, then try to fetch all storage group from config node
```

Comment:
have should be hit... and the grammar.....

Reply:
Fixed~